### DISCUSSION TITLE
Is it normal that images fetched from S3 is very slow?

### DISCUSSION THREAD

**Discussion by SkullCutter02** [directus/directus#11911] [Archive]
I configured Directus to use S3 and Cloudfront as its file storage, but it takes a long time to fetch images. A 500B image takes 500ms to fetch. I wonder if this is a directus problem, or is it an S3 problem?

**Comment by rijkvanzanten** [comment]
Very hard to tell without more information around the exact setup you're using. Could be a latency thing between where Directus is running, and where the S3 bucket is located for example

**Reply by SkullCutter02** [reply]
Directus is hosted on Heroku using a free dyno. The S3 is located at us-east-2, but I have set up a cloudfront so the location shouldn't really matter.

**Reply by rijkvanzanten** [reply]
Directus will directly connect to S3 to stream the file, so it can do permissions checking and asset generation on the fly. I'm assuming that the connection between Heroku and s3 on us-east-2 is slow, and that heroku's free dyno is a little underpowered when it comes to streaming files 🤔

**Reply by bereikme** [reply]
@rijkvanzanten Hi there Rijk. I've been struggling with this myself as well. Though I don't use S3, even on a Directus install at Digital Ocean it's slow.

**Reply by rijkvanzanten** [reply]
@bereikme could you elaborate on the exact setup? Directus install on DigitalOcean, using what exactly as file storage? DigitalOcean Spaces? Are you using the same region for spaces as you are for Directus?

**Reply by bereikme** [reply]
Hi there Rijk! It's just a simple droplet running the app, the database and also providing storage. So, all-in-one.

**Comment by ErenPhayte** [comment]
I get the same thing. Our AWS fargate instance is in the same territory as the s3 bucket. So latency shouldn't be an issue. However, when calling an asset API endpoint, it takes incredibly long, even with small images. It then times out, and if you refresh, it serves the image almost immediately.

**Reply by rijkvanzanten** [reply]
Ahh gotcha. Yes, the first time you request a generated assed (like new thumbnails or generated sizes), the asset is generated on the fly, and then returned. Subsequent calls for the same asset size will come straight from the file system 👍🏻

**Reply by bereikme** [reply]
Would it be possible to automatically generate thumbnails while saving the image?

**Reply by rijkvanzanten** [reply]
That's not currently a thing, as you can have an infinite number of thumbnail size permutations. We could look into an improvement that pre-generates the configured presets on upload though 👍🏻

**Reply by ErenPhayte** [reply]
Many other systems allow you to set up the size(s) required for that image via the CMS, so when you save a record, it will save to those sizes only. This gives us more control over how images are sized. This will then further improve the frontend calls as the images are sized already. In addition, v9 is missing the CloudFront configurations (v8 had this). This is super important when we need the images to be served via CDN and not directly from Directus.

**Reply by benhaynes** [reply]
We do have settings to limit thumbnails to a subset of options:

https://docs.directus.io/reference/files/#requesting-a-thumbnail
https://docs.directus.io/configuration/project-settings/#files-storage

**Reply by ErenPhayte** [reply]
yeah we using those options but still very slow. Perhaps the solution could be to automatically create fixed sizes based on that list to pre-warm it so those images, when called, serve very quickly instead of applying transformations on the fly (or you can move this control to each image uploader field as they could be different for different content sections) - everything else that is dynamically sized on the frontend works the way it currently does.

**Reply by benhaynes** [reply]
That could get built-in, but might be taxing on the server depending on count/complexity. Our upcoming Data Flows feature would enable this though:

1. Trigger on image upload
2. Call image presets by URL to generate (on the fly)
3. There is also an interval action, so perhaps batching over time would work too

**Reply by ErenPhayte** [reply]
Not sure what the Data Flows feature is but if you offload the sizing to a queue via Redis or something. You can control the batch processing to limit the amount of pressure on the system. Build in the ability to set up queue workers from Directus (like Laravel has in PHP). This could allow us to even scale this over multiple servers, where the workers live on a separate servers.

**Reply by ErenPhayte** [reply]
just found the info on Data flows - nice feature!
